jax/_src/numpy/lax_numpy.py

```diff
@@ -3821,15 +3821,30 @@ def dot(a, b, *, precision=None):  # pylint: disable=missing-docstring
 
 
 @_wraps(np.matmul, lax_description=_PRECISION_DOC)
-def matmul(a, b, *, precision=None):  # pylint: disable=missing-docstring
+def matmul(a, b, *, casting="same_kind", dtype=None, precision=None):  # pylint: disable=missing-docstring
   _check_arraylike("matmul", a, b)
   for i, x in enumerate((a, b)):
     if ndim(x) < 1:
       msg = (f"matmul input operand {i} must have ndim at least 1, "
              f"but it has ndim {ndim(x)}")
       raise ValueError(msg)
 
-  a, b = _promote_dtypes(a, b)
+  preferred_element_type = None
+  if dtype is None:
+    a, b = _promote_dtypes(a, b)
+  else:
+    lax._check_user_dtype_supported(dtype, "matmul")
+    dtype = dtypes.canonicalize_dtype(dtype)
+    promoted_dtype, _ = dtypes._lattice_result_type(a, b)
+    promoted_dtype = dtypes.canonicalize_dtype(promoted_dtype)
```

Comment:
You can replace these two lines with `promoted_dtype = result_type(a, b)`

Reply:
Great! That's much more readable.